### ISSUE TITLE
Miscalculation of 3**21 when IVSIZE is 4 and NV is DoubleDouble (long double)

### ISSUE THREAD

**Issue by sisyphus** [Perl/perl5#19625]
**Description**

<code>3 ** 21</code> is calculated to be a non-integer value slightly less than <code>3 ** 21</code>.
This happens when and only when $Config{ivsize} is 4 && $Config{nvtype} is long double && $Config{longdblkind} is in the range 5 to 8 (inclusive).
Note: I have only $Config{longdblkind} of 6. I haven't actually tested this for the cases where it's 5, 7 or 8.

**Steps to Reproduce**

$ perl -le 'print int(3 ** 21);'
10460353202

**Expected behavior**

We expect the output to be 10460353203 (== 3 ** 21).
The following one-liner shows that the more significant double is being correctly set to 10460353203 (<code>42037be295980000</code>), but the less significant double (which should be zero) has in fact been set to a small negative value (<code>bb20000000000000</code>).
```
$  perl -le 'print unpack "H*", pack "D>", 3 ** 21;'
42037be295980000bb20000000000000
```
I've checked that, in C, <code>powl(3.0, 21.0);</code> produces the correct result - and it does.
I've also checked that a value of <code>10460353203</code> is assigned correctly - and it is.

But this might yet turn out to be some bug in the underlying C implementation, rather than a perl bug.
I need to look at the relevant section of perl source, to see exactly what's happening when <code>3 ** 21</code> is evaluated.
Where might I find that "relevant section" ?

**Perl configuration**

```
$ perl -V
Summary of my perl5 (revision 5 version 35 subversion 11) configuration:
  Commit id: 83c93d52168a17129173de239a4ff05df85c6666
  Platform:
    osname=linux
    osvers=3.2.0-4-powerpc64
    archname=ppc64-linux-ld
    uname='linux debian-sis 3.2.0-4-powerpc64 #1 smp debian 3.2.78-1 ppc64 gnulinux '
    config_args='-des -Dcc=gcc -m32 -Dprefix=/home/sisyphus-sis/perl53511-m32-ld -Dusedevel -Uversiononly -Duselongdouble'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=undef
    use64bitall=undef
    uselongdouble=define
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='gcc -m32'
    ccflags ='-std=gnu99 -fwrapv -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O1'
    cppflags='-std=gnu99 -fwrapv -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion=''
    gccversion='4.6.3'
    gccosandvers=''
    intsize=4
    longsize=4
    ptrsize=4
    doublesize=8
    byteorder=4321
    doublekind=4
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=6
    ivtype='long'
    ivsize=4
    nvtype='long double'
    nvsize=16
    Off_t='off_t'
    lseeksize=8
    alignbytes=16
    prototype=define
  Linker and Libraries:
    ld='gcc -m32'
    ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/powerpc-linux-gnu /usr/lib /lib/powerpc-linux-gnu /lib/../lib /usr/lib/../lib /lib /lib64 /usr/lib64
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.13.so
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.13'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O1 -L/usr/local/lib -fstack-protector'


Characteristics of this binary (from libperl):
  Compile-time options:
    HAS_TIMES
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_DEVEL
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_LONG_DOUBLE
    USE_PERLIO
    USE_PERL_ATOF
  Built under linux
  Compiled at Apr 13 2022 11:39:02
  @INC:
    /home/sisyphus-sis/perl53511-m32-ld/lib/site_perl/5.35.11/ppc64-linux-ld
    /home/sisyphus-sis/perl53511-m32-ld/lib/site_perl/5.35.11
    /home/sisyphus-sis/perl53511-m32-ld/lib/5.35.11/ppc64-linux-ld
    /home/sisyphus-sis/perl53511-m32-ld/lib/5.35.11

```

**Comment by hvds** [comment]
> I need to look at the relevant section of perl source, to see exactly what's happening when `3 ** 21` is evaluated.
> Where might I find that "relevant section" ?

As a general rule:
```
% perl -Dt -e '3**21'
(-e:1)  const(IV(3))
(-e:1)  const(IV(21))
(-e:1)  pow

EXECUTING...

(-e:0)  enter
(-e:0)  nextstate
(-e:1)  leave
```
.. so the opcode is called "pow", and you should expect to find a function called `pp_pow` declared with the `PP(...)` macro in one of the pp*.c files, in this case pp.c:1060.

**Comment by demerphq** [comment]
On Fri, 15 Apr 2022 at 05:51, Hugo van der Sanden ***@***.***>
wrote:

> I need to look at the relevant section of perl source, to see exactly
> what's happening when 3 ** 21 is evaluated.
> Where might I find that "relevant section" ?
>
> As a general rule:
>
> perl -Dt -e '3**21'
> (-e:1)  const(IV(3))
> (-e:1)  const(IV(21))
> (-e:1)  pow
>
> EXECUTING...
>
> (-e:0)  enter
> (-e:0)  nextstate
> (-e:1)  leave
>
> .. so the opcode is called "pow", and you should expect to find a function
> called pp_pow declared with the PP(...) macro in one of the pp*.c files,
> in this case pp.c:1060.
>

That is cool. I didn't know about that option. Thanks. FWIW, i added a
print and you can see a little more:

$ ./perl -Ilib -Dt -e'print 3**21'
(-e:1) const(IV(3))
(-e:1) const(IV(21))
(-e:1) pow

EXECUTING...

(-e:0) enter
(-e:0) nextstate
(-e:1) pushmark
(-e:1) const(IV(10460353203))
(-e:1) print
(-e:1) leave

Eg, the expression is being evaluated at compile time and then stored as a
const to be used later.

Yves
-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by sisyphus** [comment]
Thanks for the tip on how to track it down to <code>pp_pow()</code>, @hvds .
Turns out it's just a buggy powl() implementation in libm - and I should have worked that out without even looking at pp_pow().

The bug comes in to play when the value returned by <code>Perl_pow(left, right)</code> ought to be an integer in the 31-bit to 34-bit range.
It should be fairly east to work around it inside Perl_pow() unless other bugs interfere.

I'll see what I can come up with, and report back once I've run it through some additional testing.
Then we can work out what, if anything, should be done.
I don't see a need to come up with a solution *immediately*.

Amazingly, this bug causes only one test in the perl test suite to fail - namely the <code>"nan + getpayload 3 ** 21"</code> test in ext/POSIX/t/math.t.
There are some interesting comments (by @jhi ?) about the same test failing on some Solaris systems. According to those comments, the failure on Solaris comes via an entirely different process, but I'm unable to check on that.
At least, with the doubledouble, the get/setpayload stuff is working correctly - it's just that <code>3 ** 21</code> is being set incorrectly.

Cheers,
Rob

**Comment by sisyphus** [comment]
For the record, here's the patch to pp.c that I'll be using in my own builds of perl:
```
--- pp.c_pp_c   2022-03-13 04:30:12 +1100
+++ pp.c        2022-04-22 12:25:21 +1000
@@ -1206,6 +1206,20 @@
         } else {
             SETn( Perl_pow( left, right) );
         }
+#elif IVSIZE == 4 && defined(LONGDOUBLE_DOUBLEDOUBLE) && defined(USE_LONG_DOUBLE)
+    /*
+    Under these conditions, if a known libm bug exists, Perl_pow() could return
+    an incorrect value if the correct value is an integer in the range of around
+    25 or more bits. The error is always quite small, so we work around it by
+    rounding to the nearest integer value ... but only if is_int is true.
+    See https://github.com/Perl/perl5/issues/19625.
+    */
+
+        if (is_int) {
+            SETn( roundl( Perl_pow( left, right) ) );
+        }
+        else SETn( Perl_pow( left, right) );
+
 #else
         SETn( Perl_pow( left, right) );
 #endif  /* HAS_AIX_POWL_NEG_BASE_BUG */

```
I found quite a few cases where where an integer raised to an integer power was returning a non-integer value.
This patch fixes that behaviour such that the correct integer value is returned.

However, I'm reticent to have it applied to the official perl source.

Firstly, I'm not sure that roundl() is guaranteed to exist. I initially tried using Perl_modfl(), but it was not even able to detect those cases where the problem arises.

Secondly, I don't know if this is even an issue on any other system where NV is DoubleDouble.

Thirdly, there are still lots of instances where an integer raised to an integer power is returning a result that's off by up to a few ULPs. (In such cases, the result is an integer ... just not quite the right one.)
Basically, this patch just fixes a few drops in the ocean.

Incidentally, I did try removing the "IVSIZE=4" restriction from the condition, but that didn't really help anything as regards the 64-bit builds.
It fixed some incorrect results, but then created nearly as many new incorrect results.
The actual problem of returning non-integer results did not occur when IVSIZE was 8. It was just that some returns are off by a few ULPs.

So I'll close this, with no further action planned.
If anyone thinks I should turn into this into a PR, just let me know and I'll oblige.

**Comment by demerphq** [comment]
On Sun, 24 Apr 2022 at 04:32, sisyphus ***@***.***> wrote:

> For the record, here's the patch to pp.c that I'll be using in my own
> builds of perl:
>
> --- pp.c_pp_c   2022-03-13 04:30:12 +1100
> +++ pp.c        2022-04-22 12:25:21 +1000
> @@ -1206,6 +1206,20 @@
>          } else {
>              SETn( Perl_pow( left, right) );
>          }
> +#elif IVSIZE == 4 && defined(LONGDOUBLE_DOUBLEDOUBLE) && defined(USE_LONG_DOUBLE)
> +    /*
> +    Under these conditions, if a known libm bug exists, Perl_pow() could return
> +    an incorrect value if the correct value is an integer in the range of around
> +    25 or more bits. The error is always quite small, so we work around it by
> +    rounding to the nearest integer value ... but only if is_int is true.
> +    See https://github.com/Perl/perl5/issues/19625.
> +    */
> +
> +        if (is_int) {
> +            SETn( roundl( Perl_pow( left, right) ) );
> +        }
> +        else SETn( Perl_pow( left, right) );
> +
>  #else
>          SETn( Perl_pow( left, right) );
>  #endif  /* HAS_AIX_POWL_NEG_BASE_BUG */
>
>
> I found quite a few cases where where an integer raised to an integer
> power was returning a non-integer value.
> This patch fixes that behaviour such that the correct integer value is
> returned.
>
> However, I'm reticent to have it applied to the official perl source.
>
> Firstly, I'm not sure that roundl() is guaranteed to exist. I initially
> tried using Perl_modfl(), but it was not even able to detect those cases
> where the problem arises.
>
> Secondly, I don't know if this is even an issue on any other system where
> NV is DoubleDouble.
>
> Thirdly, there are still lots of instances where an integer raised to an
> integer power is returning a result that's off by up to a few ULPs. (In
> such cases, the result is an integer ... just not quite the right one.)
> Basically, this patch just fixes a few drops in the ocean.
>
Is there any chance we could provide a better modfl() or fix Perl_pow?


> Incidentally, I did try removing the "IVSIZE=4" restriction from the
> condition, but that didn't really help anything as regards the 64-bit
> builds.
> It fixed some incorrect results, but then created nearly as many new
> incorrect results.
> The actual problem of returning non-integer results did not occur when
> IVSIZE was 8. It was just that some returns are off by a few ULPs.
>
> So I'll close this, with no further action planned.
> If anyone thinks I should turn into this into a PR, just let me know and
> I'll oblige.
>
I feel like we should have a PR with it as a proper patch, at the very
least we could point at it and people could apply it.

Patches in issues can be a pain to work with.

cheers,
Yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by sisyphus** [comment]
> Is there any chance we could provide a better modfl() or fix Perl_pow?
You would have to rewrite them from scratch. I don't think there's any other way.
Probably start with the libm implementation of powl (which Perl_pow currently wraps) and correct it where needed.
This is not something that I would undertake, nor would I expect anyone else to undertake it.

Whenever I can't tolerate a few ULPs of variance, I'll just use the mpfr library (via Math::MPFR) to provide correctly calculated and rounded values.

I'll create that  PR in the next day or two - though I expect that my own powerpc64 machine is the only computer on which it will ever be installed && be in any way useful.
(Still ... one never knows ;-)

Cheers,
Rob
